bite-size-notes/week-4-e-stochastic bellman equation.md

```diff
@@ -0,0 +1,167 @@
+
+
+<img width="500" alt="3" src="https://github.com/YJryu97/AA548spr2024/assets/115070210/df1e077f-ed84-4b7c-afce-edf144ecfd9f">
+
+
+
+The DP algorithm for the stochastic finite horizon optimal control problem has a similar form to its deterministic version, and shares several of its major characteristics:
+- Using tail subproblems to break down the minimization over multiple stages to single stage minimizations.
+- Generating __backwards__ for all $k$ and $x_k$ the values ${V*}_k(x_k)$, which give the optimal cost-to-go starting at stage $k$ at state $x_k$
+- Obtaining an optimal policy by minimization in the DP equations.
+- A structure that is suitable for approximation in value space, whereby we replace $V_k^\*$ by approximations $\bar{V}_k^\*$, and obtain a suboptimal policy by the corresponding minimization.
+
+### DP Algorithm for Stochastic Finite Horizon Problems
+- Start with
+
+$$
+V^*_N (x_N ) = g_N(x_N)
+$$
+
+and for $k = 0, . . . , N−1$, let
+
+$$
+  V_k^{\*} = \min_{u_k \in U_k (x_k)} \mathbb{E} \[ g_k(x_k, u_k, w_k) + V^*_{k+1}(f_k(x_k, u_k, w_k)) \]
+$$
+
+- If $u^∗_k = \mu^∗_k(x_k)$ minimizes the right side of this equation for each $x_k$ and $k$, the policy ${\pi}^{∗} = {\mu_0^{\*}, . . . , \mu_{N-1}^\*}$ is optimal.
+
+$$
+    {\mu_k} (x_k) \in arg\min_{u_k \in U_k (x_k)} \mathbb{E} \[ g_k(x_k, u_k, w_k) + V^*_{k+1}(f_k(x_k, u_k, w_k))
+$$
+
+- For finding suboptimal policy, approximate the cost-go-function $\bar{V}^*_{k}$(using Model Approximation, Neural-Nets, Rollout algorithm, Model Predictive Control..etc)
+
+$$
+   \bar{V_k}^{\*} = \min_{u_k \in U_k (x_k)} \mathbb{E} \[ g_k(x_k, u_k, w_k) + \bar{V}^*_{k+1}(f_k(x_k, u_k, w_k)) \
+$$
+
+where suboptimal control policy obtained 
+
+$$
+    \bar{\mu_k} (x_k) \in arg\min_{u_k \in U_k (x_k)} \mathbb{E} \[ g_k(x_k, u_k, w_k) + \bar{V}^*_{k+1}(f_k(x_k, u_k, w_k))
+$$
+
+
+The key fact is that for every initial state $x_0$, the optimal cost $J^\*(x_0)$ is equal to the function $J_0^\*(x_0)$, obtained at the __last step__ of the above DP algorithm.
+
+
+
+However, the DP algorithm can be very time-consuming, but more than the deterministic case since it involves the expected value operation and a rapid increase of the required computation and memory storage as the problem’s size increases. This difficulties motivate suboptimal control schemes, such as approximation in value space whereby we replace $J_k^\*$ with easier obtainable approximation $\bar{J}_k$. $\bar{J}_k$ should be “close” to the optimal cost-to-go function $J_k^\*$ for all $k$.
+schemes. From the looking ahead minimization of cost function, we obtain the suboptimal(heuristic) policies $\bar{\mu}$ which are used in situations where finding an optimal solution is computationally intractable or impractical. 
+
+## Conclusion
+During the execution of a controlled system, at any point in time the system trajectory may be split in two. 

```

Comment:
Does the idea of optimize expect value imply, their is control input that might have smaller cost, because the policy include the probability. Take hw2 as example, since we optimize the policy, but there is a chance that we don't maintain the machine and it works for all 4 weeks.

Reply:
Exactly not the policy includes the probability. For example, in scheme of model predictive control, we know that tracking all the trajectory is intractable and inefficient, hence implementing short prediction horizon for finding feasible control. However, in stochastic case, even from searching optimal policy from one-look ahead state optimal cost is still heavy because probabilistic problem contains state uncertainty and finding optimal control from added dimension of uncertainty makes the policy space is also heavy. So, we add the constraints and approximation to the problem to make the problem space more tractable but suboptimal.   Look on the line 138 which is the techniques to evade this dimensionality problem.  I'll add more description on that.